File: Leetcode_200.py
class Solution:
    def numIslands(self, grid):

        if not grid or not grid[0]:
            return 0                     # → Important: handle empty grid

        count = 0                        # → Number of islands
        n = len(grid)                    # → rows
        m = len(grid[0])                 # → cols

        def dfs(i, j):
            if i < 0 or i >= n or j < 0 or j >= m or grid[i][j] == '0':
                return                   # → Out of bounds / water / visited

            grid[i][j] = '0'             # → Mark visited (sink land)

            dfs(i + 1, j)                # down
            dfs(i - 1, j)                # up
            dfs(i, j + 1)                # right
            dfs(i, j - 1)                # left

        for i in range(n):               # → O(n × m)
            for j in range(m):
                if grid[i][j] == '1':    # → Found new island
                    dfs(i, j)            # → Flood fill
                    count += 1

        return count                     # → Total islands

File: Leetcode_200.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {

        if (grid.empty() || grid[0].empty())
            return 0;                               // → Important: handle empty grid

        int count = 0;                              // → Number of islands
        int n = grid.size();                        // → rows
        int m = grid[0].size();                     // → cols

        for (int i = 0; i < n; i++) {               // → O(n × m)
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1') {            // → Found land → new island
                    dfs(grid, i, j);                // → Flood fill the island
                    count++;                        // → Increase island count
                }
            }
        }

        return count;                               // → Total islands
    }

private:
    void dfs(vector<vector<char>>& grid, int i, int j) {

        if (i < 0 || i >= grid.size() ||            // → Out of bounds
            j < 0 || j >= grid[0].size() ||
            grid[i][j] == '0')
            return;                                 // → Already water/visited

        grid[i][j] = '0';                           // → Mark visited (sink land)

        dfs(grid, i + 1, j);                        // down
        dfs(grid, i - 1, j);                        // up
        dfs(grid, i, j + 1);                        // right
        dfs(grid, i, j - 1);                        // left
    }
};
